### PR TITLE
ДЗ Спринт #27

### PR DESCRIPTION
**Pull request by mr-friderick**
Выполнял Уровень 1. Замена UI нескольких экранов на Compose с сохранением традиционного подхода к навигации.

Неиспользуемые фрагменты пока что не удалял. Если осилю у будущем перевести все на compose - почищу.
В некоторых модулях слоев Data и Domain поменял типы используемых значений, привел к однотипности (ранее использовался ArrayList, перевел просто на List)

### CODE REVIEW

File: app/src/main/java/com/example/playlistmaker/medialibrary/ui/compose/FavoritesTracksScreen.kt [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+    tracks: List<Track>,
+    onTrackClick: (Track) -> Unit
+) {
+    LazyColumn(
+        modifier = Modifier
+            .fillMaxSize()
+            .padding(top = 24.dp)
+            .padding(bottom = 20.dp)
+            .padding(start = 13.dp)
+            .padding(end = 20.dp),
```
Comment (ArturNurtdinov) [inline app/src/main/java/com/example/playlistmaker/medialibrary/ui/compose/FavoritesTracksScreen.kt]: Можно один раз `.padding` написать и в одном вызове проставить все 4 отступа

File: app/src/main/java/com/example/playlistmaker/medialibrary/ui/fragments/MediaRootFragment.kt [hunk 2]
```diff
@@ -4,48 +4,75 @@ import android.os.Bundle
...
-    override fun onDestroyView() {
-        super.onDestroyView()
-        tabMediator.detach()
-    }
-
-    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
-        super.onViewCreated(view, savedInstanceState)
-
-        initVariables()
+        return ComposeView(requireContext()).apply {
```
Comment (ArturNurtdinov) [inline app/src/main/java/com/example/playlistmaker/medialibrary/ui/fragments/MediaRootFragment.kt]: Во фрагментах лучше поменять `ViewCompositionStrategy` - `setViewCompositionStrategy(ViewCompositionStrategy.DisposeOnViewTreeLifecycleDestroyed)`, тогда композиция уничтожалась будет уничтожена вместе с вью. Поведение по умолчанию уничтожает композицию при каждом отсоединении вью от Window, что происходит чаще, чем полное уничтожение вью

File: app/src/main/java/com/example/playlistmaker/settings/ui/compose/SettingsCompose.kt [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+    val context = LocalContext.current
+    val isChecked by viewModel.themeLiveData.observeAsState(false)
+
+    val dataForIntent: MutableMap<String, String> = mutableMapOf()
+    dataForIntent["shareLink"] = stringResource(R.string.settings_share_link)
+    dataForIntent["shareTitle"] = stringResource(R.string.settings_share_title)
+    dataForIntent["supportEmail"] = stringResource(R.string.settings_support_email)
+    dataForIntent["supportSubject"] = stringResource(R.string.settings_support_subject)
+    dataForIntent["supportMessage"] = stringResource(R.string.settings_support_message)
+    dataForIntent["agreementLink"] = stringResource(R.string.settings_agreement_link)
```
Comment (ArturNurtdinov) [inline app/src/main/java/com/example/playlistmaker/settings/ui/compose/SettingsCompose.kt]: Не стоит в композабл функции собирать целую мапу, т.к. создание объектов в композиции плохо влияет на перформанс. Здесь лучше хотя бы в `remember` обернуть создание мапы

File: app/build.gradle.kts [hunk 4]
```diff
@@ -81,4 +85,18 @@ dependencies {
...
+    implementation(libs.androidx.compose.ui.graphics)
+    implementation(libs.androidx.compose.ui.tooling.preview)
+    implementation(libs.androidx.compose.material3)
+
+    androidTestImplementation(platform(libs.androidx.compose.bom))
+    androidTestImplementation(libs.androidx.compose.ui.test.junit4)
+    debugImplementation(libs.androidx.compose.ui.tooling)
+    debugImplementation(libs.androidx.compose.ui.test.manifest)
+
+    implementation("com.github.skydoves:landscapist-glide:2.6.1")
```
Comment (ArturNurtdinov) [inline app/build.gradle.kts]: В компоузе вместо глайда лучше использовать Coil, т.к. глайд пока не мигрировали под компоуз

File: app/src/main/java/com/example/playlistmaker/medialibrary/ui/compose/FavoritesTracksScreen.kt [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+        viewModel.setContent()
+    }
+
+    when (tracksViewState) {
+        is FavoriteTracksViewState.Content -> {
+            Content(
+                tracks = (tracksViewState as FavoriteTracksViewState.Content).favoriteTracks,
+                onTrackClick = openPlayer
+            )
+        }
```
Comment (ArturNurtdinov) [inline app/src/main/java/com/example/playlistmaker/medialibrary/ui/compose/FavoritesTracksScreen.kt]: Тут можно от каста избавиться следующим образом:
```suggestion
    when (val uiState = tracksViewState) {
        is FavoriteTracksViewState.Content -> {
            Content(
                tracks = uiState.favoriteTracks,
                onTrackClick = openPlayer
            )
        }
```

File: app/src/main/java/com/example/playlistmaker/medialibrary/ui/compose/ListPlaylistScreen.kt [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+@SuppressLint("LocalContextResourcesRead")
+@Composable
+private fun ItemContent(
+    playlist: Playlist,
+    onClickListener: () -> Unit
+) {
+
+    val context = LocalContext.current
+    val posterFile = File(context.filesDir, playlist.picturePath)
+    val tracksCount = context.resources.getQuantityString(
```
Comment (ArturNurtdinov) [inline app/src/main/java/com/example/playlistmaker/medialibrary/ui/compose/ListPlaylistScreen.kt]: Для компоуза есть метод `pluralStringResource`, лучше использовать его чтобы плюралс вытащить

**Review by ArturNurtdinov — COMMENTED** [review]
Отсутствует плейхсолдер обложки плейлиста на экране плейлистов:

image

Макет:

image

File: app/src/main/java/com/example/playlistmaker/medialibrary/ui/compose/FavoritesTracksScreen.kt (hunk 1 above)
Comment (mr-friderick) [reply app/src/main/java/com/example/playlistmaker/medialibrary/ui/compose/FavoritesTracksScreen.kt]: Спасибо, поправил

File: app/src/main/java/com/example/playlistmaker/medialibrary/ui/fragments/MediaRootFragment.kt (hunk 2 above)
Comment (mr-friderick) [reply app/src/main/java/com/example/playlistmaker/medialibrary/ui/fragments/MediaRootFragment.kt]: Надеюсь понял о чем речь, поправил. Доп. почитаю об этом

File: app/src/main/java/com/example/playlistmaker/settings/ui/compose/SettingsCompose.kt (hunk 3 above)
Comment (mr-friderick) [reply app/src/main/java/com/example/playlistmaker/settings/ui/compose/SettingsCompose.kt]: Понял, поправил через remember

File: app/src/main/java/com/example/playlistmaker/medialibrary/ui/compose/FavoritesTracksScreen.kt (hunk 5 above)
Comment (mr-friderick) [reply app/src/main/java/com/example/playlistmaker/medialibrary/ui/compose/FavoritesTracksScreen.kt]: Поправил

File: app/src/main/java/com/example/playlistmaker/medialibrary/ui/compose/ListPlaylistScreen.kt (hunk 6 above)
Comment (mr-friderick) [reply app/src/main/java/com/example/playlistmaker/medialibrary/ui/compose/ListPlaylistScreen.kt]: Готово

File: app/build.gradle.kts (hunk 4 above)
Comment (mr-friderick) [reply app/build.gradle.kts]: Заменил на Coil, убрал часть дублирования. Зависимость от Glide удалил.
